# ELephantSQL Database Builder:

A script for importing local aviation data; modifying it to create various tables of interest; and then importing those tables into a database so that my web app can access it

By Kenneth Burchfiel

Released under the MIT License

In [1]:
import json
import sqlalchemy
import pandas as pd



Connecting to ElephantSQL database:

In [2]:
with open ("..\key_paths\path_to_keys_folder.txt") as file:
    key_path = file.read()

In [3]:
with open(key_path+"\\elephantsql_kjb3webchartapp_db_url.txt") as file:
    db_url = file.read()
    # This code reads in my database's URL. This URL is listed on the home page for my database within elephantsql.com. As shown below, SQLAlchemy can use this URL to connect to the database. 

elephantsql_db_url_for_sqlalchemy = db_url.replace('postgres://', 'postgresql://')
# This change, which is required for SQLAlchemy to work correctly, is based on the code suggested at:
# # https://help.heroku.com/ZKNTJQSK/why-is-sqlalchemy-1-4-x-not-connecting-to-heroku-postgres

In [4]:
elephantsql_engine = sqlalchemy.create_engine(elephantsql_db_url_for_sqlalchemy)

## Loading in data

Next, I'll load in a .csv file that was created using the [flights_table_builder](https://github.com/kburchfiel/route_maps_builder/blob/master/flights_table_builder.ipynb) script within my [Route Maps Builder](https://github.com/kburchfiel/route_maps_builder) program. The original file is too large to host on GitHub, but it can be found within [this Google Drive page](https://drive.google.com/file/d/1gUozlxpVIO-gQ6d-AQ4Qgpg91Ut3kqUD/view?usp=sharing).



In [5]:
pd.set_option('display.max_columns', 100)

In [6]:
df_flights = pd.read_csv(r'C:\Users\kburc\D1V1\Documents\!Dell64docs\Programming\py\kjb3_programs\route_maps_builder\routes_planes_coordinates_for_mapping_2018.csv')
df_flights.drop('Unnamed: 0', axis = 1,inplace=True)
df_flights['UNIQUE_CARRIER_NAME'].fillna("Unknown", inplace = True) # This line will allow rows without an airline name to still be processed by pivot table operations that reference those names.
df_flights

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED    PAYLOAD   SEATS  \
0                        0.0                   1.0    21502.0    76.0   
1                        0.0                   3.0    64506.0   228.0   
2                        0.0                   1.0    21502.0    76.0   
3                        0.0                   1.0    21502.0    76.0   
4                        0.0                   1.0    21502.0    76.0   
...                      ...                   ...        ...     ...   
349168                1146.0                 573.0  1031400.0  5157.0   
349169                1166.0                 583.0  1049400.0  5247.0   
349170                1188.0                 594.0  1069200.0  5346.0   
349171                1216.0                 608.0  1094400.0  5472.0   
349172                1258.0                 629.0  1132200.0  5661.0   

        PASSENGERS  FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME  \
0              3.0      0.0   0.0     901.0         170.0     140.0   
1             75.0      0.0   0.0     228.0         219.0     140.0   
2             64.0      0.0   0.0     851.0         144.0     114.0   
3             55.0      0.0   0.0     122.0          58.0      31.0   
4              0.0      0.0   0.0     476.0         157.0     138.0   
...            ...      ...   ...       ...           ...       ...   
349168      3693.0      0.0   0.0      91.0       26816.0   20972.0   
349169      3646.0      0.0   0.0      91.0       27284.0   21338.0   
349170      3573.0      0.0   0.0      91.0       27799.0   21740.0   
349171      3827.0      0.0   0.0      91.0       28454.0   22253.0   
349172      4056.0      0.0   0.0      91.0       29437.0   23021.0   

       UNIQUE_CARRIER  AIRLINE_ID UNIQUE_CARRIER_NAME UNIQUE_CARRIER_ENTITY  \
0                  9E     20363.0   Endeavor Air Inc.                 07024   
1                  9E     20363.0   Endeavor Air Inc.                 07024   
2                  9E     20363.0   Endeavor Air Inc.                 07024   
3                  9E     20363.0   Endeavor Air Inc.                 07024   
4                  9E     20363.0   Endeavor Air Inc.                 07024   
...               ...         ...                 ...                   ...   
349168             9K     20253.0            Cape Air                 07021   
349169             9K     20253.0            Cape Air                 07021   
349170             9K     20253.0            Cape Air                 07021   
349171             9K     20253.0            Cape Air                 07021   
349172             9K     20253.0            Cape Air                 07021   

       REGION CARRIER       CARRIER_NAME  CARRIER_GROUP  CARRIER_GROUP_NEW  \
0           D      9E  Endeavor Air Inc.              2                2.0   
1           D      9E  Endeavor Air Inc.              2                2.0   
2           D      9E  Endeavor Air Inc.              2                2.0   
3           D      9E  Endeavor Air Inc.              2                2.0   
4           D      9E  Endeavor Air Inc.              2                2.0   
...       ...     ...                ...            ...                ...   
349168      D      9K           Cape Air              1                6.0   
349169      D      9K           Cape Air              1                6.0   
349170      D      9K           Cape Air              1                6.0   
349171      D      9K           Cape Air              1                6.0   
349172      D      9K           Cape Air              1                6.0   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
0                   12264                1226402                  30852   
1                   12264                1226402                  30852   
2                   12266                1226603                  31453   
3                   12323                1232305                  32323   
4                

## Pivoting and filtering data

This table currently contains around 350,000 rows and 65 columns. However, my web app doesn't need all of those rows present in its own PostgreSQL database; instead, it just needs enough data to generate various charts. Therefore, I'll now create some modified versions of this data and then upload it to the PostgreSQL database.

In [7]:
df_airline_origin_dest_plane_pivot = df_flights.pivot_table(index = ['UNIQUE_CARRIER', 'ORIGIN_DEST', 'Plane_Type_Text'], values = 'PASSENGERS', aggfunc = 'sum').reset_index().sort_values('PASSENGERS', ascending = False)
df_airline_origin_dest_plane_pivot = df_airline_origin_dest_plane_pivot.query("PASSENGERS > 100000").copy()
df_airline_origin_dest_plane_pivot['Airport 1'] = df_airline_origin_dest_plane_pivot['ORIGIN_DEST'].str.split('_').str[0]
df_airline_origin_dest_plane_pivot['Airport 2'] = df_airline_origin_dest_plane_pivot['ORIGIN_DEST'].str.split('_').str[1]
df_airline_origin_dest_plane_pivot.rename(columns={"UNIQUE_CARRIER":"Airline","PASSENGERS":"Passengers","ORIGIN_DEST":"Origin_Dest"},inplace=True)
df_airline_origin_dest_plane_pivot.reset_index(drop=True,inplace=True)
df_airline_origin_dest_plane_pivot

Airline Origin_Dest                Plane_Type_Text  Passengers Airport 1  \
0         HA     HNL_OGG                 Boeing 717-200   1954139.0       HNL   
1         DL     ATL_MCO                 Boeing 757-200   1417832.0       ATL   
2         HA     HNL_KOA                 Boeing 717-200   1281221.0       HNL   
3         WN     DAL_HOU     Boeing 737-700/700LR/Max 7   1260362.0       DAL   
4         AA     DFW_LAX       Airbus Industrie A321/Lr   1257512.0       DFW   
...      ...         ...                            ...         ...       ...   
2799      NK     FLL_LGA  Airbus Industrie A320-100/200    100381.0       FLL   
2800      EI     DUB_MCO      Airbus Industrie A330-200    100329.0       DUB   
2801      YX     DCA_MCI                Embraer ERJ-175    100303.0       DCA   
2802      WN     BNA_CLT     Boeing 737-700/700LR/Max 7    100282.0       BNA   
2803      EV     BHM_IAH                    Embraer-145    100014.0       BHM   

     Airport 2  
0          OGG  
1          MCO  
2          KOA  
3          HOU  
4          LAX  
...        ...  
2799       LGA  
2800       MCO  
2801       MCI  
2802       CLT  
2803       IAH  

[2804 rows x 6 columns]

The following cell exports this pivot table to .csv format locally (so that it can be accessed without a database connection) and to .sql format online.

In [8]:
df_airline_origin_dest_plane_pivot.to_csv('local_copy_of_airports_airlines_aircraft_2018.csv', index = False)

df_airline_origin_dest_plane_pivot.to_sql("airports_airlines_aircraft_2018",if_exists = 'replace', con = elephantsql_engine, index = False)

804

I believe the output of the above cell shows the smallest 3 digits of the number of rows updated within the table. Thus, if 2,804 rows were created within the table, the output will be 804.

The following cell reads the new database table back into Pandas in order to verify that it was imported successfully:

In [9]:
pd.read_sql("airports_airlines_aircraft_2018", con = elephantsql_engine)

Airline Origin_Dest                Plane_Type_Text  Passengers Airport 1  \
0         HA     HNL_OGG                 Boeing 717-200   1954139.0       HNL   
1         DL     ATL_MCO                 Boeing 757-200   1417832.0       ATL   
2         HA     HNL_KOA                 Boeing 717-200   1281221.0       HNL   
3         WN     DAL_HOU     Boeing 737-700/700LR/Max 7   1260362.0       DAL   
4         AA     DFW_LAX       Airbus Industrie A321/Lr   1257512.0       DFW   
...      ...         ...                            ...         ...       ...   
2799      NK     FLL_LGA  Airbus Industrie A320-100/200    100381.0       FLL   
2800      EI     DUB_MCO      Airbus Industrie A330-200    100329.0       DUB   
2801      YX     DCA_MCI                Embraer ERJ-175    100303.0       DCA   
2802      WN     BNA_CLT     Boeing 737-700/700LR/Max 7    100282.0       BNA   
2803      EV     BHM_IAH                    Embraer-145    100014.0       BHM   

     Airport 2  
0          OGG  
1          MCO  
2          KOA  
3          HOU  
4          LAX  
...        ...  
2799       LGA  
2800       MCO  
2801       MCI  
2802       CLT  
2803       IAH  

[2804 rows x 6 columns]

## Creating additional tables for the database:

The following cells create filtered versions of df_flights that show only domestic or international destinations. This allows me to ensure, for instance, that a list of the top 20 US airports by international traffic won't contain non-US airports.

In [10]:
df_flights_domestic_destination_airports = df_flights.query("Destination_Region == 'Domestic'")
df_flights_domestic_destination_airports

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED    PAYLOAD   SEATS  \
0                        0.0                   1.0    21502.0    76.0   
1                        0.0                   3.0    64506.0   228.0   
2                        0.0                   1.0    21502.0    76.0   
3                        0.0                   1.0    21502.0    76.0   
4                        0.0                   1.0    21502.0    76.0   
...                      ...                   ...        ...     ...   
349168                1146.0                 573.0  1031400.0  5157.0   
349169                1166.0                 583.0  1049400.0  5247.0   
349170                1188.0                 594.0  1069200.0  5346.0   
349171                1216.0                 608.0  1094400.0  5472.0   
349172                1258.0                 629.0  1132200.0  5661.0   

        PASSENGERS  FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME  \
0              3.0      0.0   0.0     901.0         170.0     140.0   
1             75.0      0.0   0.0     228.0         219.0     140.0   
2             64.0      0.0   0.0     851.0         144.0     114.0   
3             55.0      0.0   0.0     122.0          58.0      31.0   
4              0.0      0.0   0.0     476.0         157.0     138.0   
...            ...      ...   ...       ...           ...       ...   
349168      3693.0      0.0   0.0      91.0       26816.0   20972.0   
349169      3646.0      0.0   0.0      91.0       27284.0   21338.0   
349170      3573.0      0.0   0.0      91.0       27799.0   21740.0   
349171      3827.0      0.0   0.0      91.0       28454.0   22253.0   
349172      4056.0      0.0   0.0      91.0       29437.0   23021.0   

       UNIQUE_CARRIER  AIRLINE_ID UNIQUE_CARRIER_NAME UNIQUE_CARRIER_ENTITY  \
0                  9E     20363.0   Endeavor Air Inc.                 07024   
1                  9E     20363.0   Endeavor Air Inc.                 07024   
2                  9E     20363.0   Endeavor Air Inc.                 07024   
3                  9E     20363.0   Endeavor Air Inc.                 07024   
4                  9E     20363.0   Endeavor Air Inc.                 07024   
...               ...         ...                 ...                   ...   
349168             9K     20253.0            Cape Air                 07021   
349169             9K     20253.0            Cape Air                 07021   
349170             9K     20253.0            Cape Air                 07021   
349171             9K     20253.0            Cape Air                 07021   
349172             9K     20253.0            Cape Air                 07021   

       REGION CARRIER       CARRIER_NAME  CARRIER_GROUP  CARRIER_GROUP_NEW  \
0           D      9E  Endeavor Air Inc.              2                2.0   
1           D      9E  Endeavor Air Inc.              2                2.0   
2           D      9E  Endeavor Air Inc.              2                2.0   
3           D      9E  Endeavor Air Inc.              2                2.0   
4           D      9E  Endeavor Air Inc.              2                2.0   
...       ...     ...                ...            ...                ...   
349168      D      9K           Cape Air              1                6.0   
349169      D      9K           Cape Air              1                6.0   
349170      D      9K           Cape Air              1                6.0   
349171      D      9K           Cape Air              1                6.0   
349172      D      9K           Cape Air              1                6.0   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
0                   12264                1226402                  30852   
1                   12264                1226402                  30852   
2                   12266                1226603                  31453   
3                   12323                1232305                  32323   
4                

In [11]:
df_flights_intl_destination_airports = df_flights.query("Destination_Region == 'International'")
df_flights_intl_destination_airports

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED   PAYLOAD   SEATS  \
33                       0.0                   1.0   21502.0    76.0   
89                       0.0                   1.0   21405.0     0.0   
90                       0.0                   1.0   21323.0     0.0   
91                       0.0                   2.0   42143.0     0.0   
92                       0.0                   1.0   20995.0     0.0   
...                      ...                   ...       ...     ...   
347746                 246.0                 123.0  221400.0  1107.0   
348082                 262.0                 131.0  235800.0  1179.0   
348512                 292.0                 146.0  262800.0  1314.0   
348551                 296.0                 148.0  266400.0  1332.0   
348696                 322.0                 161.0  447580.0  1449.0   

        PASSENGERS  FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME  \
33            66.0      0.0   0.0     182.0          43.0      32.0   
89             0.0      0.0   0.0    1335.0         232.0     220.0   
90             0.0   5090.0   0.0     165.0          47.0      33.0   
91             0.0      0.0   0.0     565.0         200.0     170.0   
92             0.0   4069.0   0.0     878.0         148.0     127.0   
...            ...      ...   ...       ...           ...       ...   
347746       455.0      0.0   0.0      96.0        5904.0    4797.0   
348082       761.0      0.0   0.0      96.0        6288.0    5109.0   
348512       712.0      0.0   0.0      96.0        7008.0    5694.0   
348551       884.0      0.0   0.0      96.0        7104.0    5772.0   
348696       972.0      0.0   0.0      96.0        7728.0    6279.0   

       UNIQUE_CARRIER  AIRLINE_ID  UNIQUE_CARRIER_NAME UNIQUE_CARRIER_ENTITY  \
33                 9E     20363.0    Endeavor Air Inc.                 07024   
89                AMQ     20201.0  Ameristar Air Cargo                 16057   
90                AMQ     20201.0  Ameristar Air Cargo                 16057   
91                AMQ     20201.0  Ameristar Air Cargo                 16057   
92                AMQ     20201.0  Ameristar Air Cargo                 16057   
...               ...         ...                  ...                   ...   
347746             9K     20253.0             Cape Air                 07021   
348082             9K     20253.0             Cape Air                 07021   
348512             9K     20253.0             Cape Air                 07021   
348551             9K     20253.0             Cape Air                 07021   
348696             9K     20253.0             Cape Air                 07021   

       REGION CARRIER         CARRIER_NAME  CARRIER_GROUP  CARRIER_GROUP_NEW  \
33          D      9E    Endeavor Air Inc.              2                2.0   
89          I     AMQ  Ameristar Air Cargo              1                4.0   
90          I     AMQ  Ameristar Air Cargo              1                4.0   
91          I     AMQ  Ameristar Air Cargo              1                4.0   
92          I     AMQ  Ameristar Air Cargo              1                4.0   
...       ...     ...                  ...            ...                ...   
347746      D      9K             Cape Air              1                6.0   
348082      D      9K             Cape Air              1                6.0   
348512      D      9K             Cape Air              1                6.0   
348551      D      9K             Cape Air              1                6.0   
348696      D      9K             Cape Air              1                6.0   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
33                  14576                1457606                  34576   
89                  11292                1129202                  30325   
90                  13061                1306106                  33038   
91                  13061                1306106                  33

In [12]:
df_flights_domestic_destination_airports_intl_traffic = df_flights.query("Destination_Region == 'Domestic' & Route_Type == 'International'")
df_flights_domestic_destination_airports_intl_traffic

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED   PAYLOAD   SEATS  \
36                       0.0                   1.0   21502.0    76.0   
88                       0.0                   1.0   21323.0     0.0   
93                       0.0                   2.0   42400.0     0.0   
97                       0.0                   1.0   21148.0     0.0   
98                       0.0                   3.0   63313.0     0.0   
...                      ...                   ...       ...     ...   
347799                 248.0                 124.0  223200.0  1116.0   
348078                 262.0                 131.0  235800.0  1179.0   
348532                 294.0                 147.0  264600.0  1323.0   
348550                 296.0                 148.0  266400.0  1332.0   
348729                 328.0                 164.0  455920.0  1476.0   

        PASSENGERS  FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME  \
36            75.0      0.0   0.0     803.0         256.0     203.0   
88             0.0      0.0   0.0     468.0          75.0      60.0   
93             0.0      0.0   0.0     128.0          82.0      54.0   
97             0.0      0.0   0.0     376.0          80.0      60.0   
98             0.0      0.0   0.0     165.0         141.0      87.0   
...            ...      ...   ...       ...           ...       ...   
347799       663.0      0.0   0.0      96.0        5952.0    4836.0   
348078       695.0      0.0   0.0      96.0        6288.0    5109.0   
348532       805.0      0.0   0.0      96.0        7056.0    5733.0   
348550       864.0      0.0   0.0      96.0        7104.0    5772.0   
348729       716.0      0.0   0.0      96.0        7872.0    6396.0   

       UNIQUE_CARRIER  AIRLINE_ID  UNIQUE_CARRIER_NAME UNIQUE_CARRIER_ENTITY  \
36                 9E     20363.0    Endeavor Air Inc.                 07024   
88                AMQ     20201.0  Ameristar Air Cargo                 16057   
93                AMQ     20201.0  Ameristar Air Cargo                 16057   
97                AMQ     20201.0  Ameristar Air Cargo                 16057   
98                AMQ     20201.0  Ameristar Air Cargo                 16057   
...               ...         ...                  ...                   ...   
347799             9K     20253.0             Cape Air                 07021   
348078             9K     20253.0             Cape Air                 07021   
348532             9K     20253.0             Cape Air                 07021   
348550             9K     20253.0             Cape Air                 07021   
348729             9K     20253.0             Cape Air                 07021   

       REGION CARRIER         CARRIER_NAME  CARRIER_GROUP  CARRIER_GROUP_NEW  \
36          D      9E    Endeavor Air Inc.              2                2.0   
88          I     AMQ  Ameristar Air Cargo              1                4.0   
93          I     AMQ  Ameristar Air Cargo              1                4.0   
97          I     AMQ  Ameristar Air Cargo              1                4.0   
98          I     AMQ  Ameristar Air Cargo              1                4.0   
...       ...     ...                  ...            ...                ...   
347799      D      9K             Cape Air              1                6.0   
348078      D      9K             Cape Air              1                6.0   
348532      D      9K             Cape Air              1                6.0   
348550      D      9K             Cape Air              1                6.0   
348729      D      9K             Cape Air              1                6.0   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
36                  16217                1621702                  36083   
88                  10638                1063802                  30638   
93                  13514                1351403                  33514   
97                  14879                1487901                  34

# Creating a U.S. airports table

I'll now create a list of the top 20 U.S. airports by arriving passenger traffic that will also include (1) domestic and international traffic breakdowns and (2) each airport's rank. (I chose to focus on arriving traffic rather than departing traffic, but either option can work, as long as you're consistent in that decision throughout your code.) First, I'll calculate the ranks of the top 20 U.S. airports by arriving passenger traffic:

In [13]:
df_top_20_airport_ranks = df_flights_domestic_destination_airports.pivot_table(index = "DEST", values = "PASSENGERS", aggfunc = "sum").sort_values("PASSENGERS", ascending = False).head(20).reset_index()
df_top_20_airport_ranks['Rank'] = df_top_20_airport_ranks['PASSENGERS'].rank(ascending = False).astype(int) # See https://pandas.pydata.org/docs/reference/api/pandas.Series.rank.html
df_top_20_airport_ranks

DEST  PASSENGERS  Rank
0   ATL  47388875.0     1
1   LAX  41502828.0     2
2   ORD  36572207.0     3
3   DFW  30823751.0     4
4   JFK  29744941.0     5
5   DEN  29527533.0     6
6   SFO  26974833.0     7
7   LAS  23511649.0     8
8   MCO  23178413.0     9
9   SEA  22558466.0    10
10  EWR  22089153.0    11
11  PHX  21388493.0    12
12  IAH  20804509.0    13
13  MIA  20634098.0    14
14  CLT  20108513.0    15
15  BOS  19925161.0    16
16  FLL  17498368.0    17
17  MSP  17016621.0    18
18  DTW  16067119.0    19
19  PHL  14721067.0    20

Next, I'll create a list of these top 20 airports:

In [14]:
top_20_airports = df_top_20_airport_ranks['DEST'].to_list()
top_20_airports

['ATL',
 'LAX',
 'ORD',
 'DFW',
 'JFK',
 'DEN',
 'SFO',
 'LAS',
 'MCO',
 'SEA',
 'EWR',
 'PHX',
 'IAH',
 'MIA',
 'CLT',
 'BOS',
 'FLL',
 'MSP',
 'DTW',
 'PHL']

I'll then use this list to create a filtered version of df_flights:

In [15]:
df_top_20_domestic_airports = df_flights.query("DEST in @top_20_airports")
df_top_20_domestic_airports

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED    PAYLOAD   SEATS  \
0                        0.0                   1.0    21502.0    76.0   
1                        0.0                   3.0    64506.0   228.0   
11                       0.0                   2.0    43004.0   152.0   
14                       0.0                   1.0    21502.0    76.0   
15                       0.0                   2.0    43004.0   152.0   
...                      ...                   ...        ...     ...   
349161                 910.0                 455.0   819000.0  4095.0   
349165                1096.0                 548.0   986400.0  4932.0   
349168                1146.0                 573.0  1031400.0  5157.0   
349169                1166.0                 583.0  1049400.0  5247.0   
349171                1216.0                 608.0  1094400.0  5472.0   

        PASSENGERS  FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME  \
0              3.0      0.0   0.0     901.0         170.0     140.0   
1             75.0      0.0   0.0     228.0         219.0     140.0   
11           114.0      0.0   0.0     762.0         322.0     264.0   
14            67.0      0.0   0.0     304.0          80.0      57.0   
15            73.0      0.0   0.0     404.0         197.0     129.0   
...            ...      ...   ...       ...           ...       ...   
349161      2360.0      0.0   0.0      91.0       21294.0   16653.0   
349165      2913.0      0.0   0.0      91.0       25646.0   20057.0   
349168      3693.0      0.0   0.0      91.0       26816.0   20972.0   
349169      3646.0      0.0   0.0      91.0       27284.0   21338.0   
349171      3827.0      0.0   0.0      91.0       28454.0   22253.0   

       UNIQUE_CARRIER  AIRLINE_ID UNIQUE_CARRIER_NAME UNIQUE_CARRIER_ENTITY  \
0                  9E     20363.0   Endeavor Air Inc.                 07024   
1                  9E     20363.0   Endeavor Air Inc.                 07024   
11                 9E     20363.0   Endeavor Air Inc.                 07024   
14                 9E     20363.0   Endeavor Air Inc.                 07024   
15                 9E     20363.0   Endeavor Air Inc.                 07024   
...               ...         ...                 ...                   ...   
349161             9K     20253.0            Cape Air                 07021   
349165             9K     20253.0            Cape Air                 07021   
349168             9K     20253.0            Cape Air                 07021   
349169             9K     20253.0            Cape Air                 07021   
349171             9K     20253.0            Cape Air                 07021   

       REGION CARRIER       CARRIER_NAME  CARRIER_GROUP  CARRIER_GROUP_NEW  \
0           D      9E  Endeavor Air Inc.              2                2.0   
1           D      9E  Endeavor Air Inc.              2                2.0   
11          D      9E  Endeavor Air Inc.              2                2.0   
14          D      9E  Endeavor Air Inc.              2                2.0   
15          D      9E  Endeavor Air Inc.              2                2.0   
...       ...     ...                ...            ...                ...   
349161      D      9K           Cape Air              1                6.0   
349165      D      9K           Cape Air              1                6.0   
349168      D      9K           Cape Air              1                6.0   
349169      D      9K           Cape Air              1                6.0   
349171      D      9K           Cape Air              1                6.0   

        ORIGIN_AIRPORT_ID  ORIGIN_AIRPORT_SEQ_ID  ORIGIN_CITY_MARKET_ID  \
0                   12264                1226402                  30852   
1                   12264                1226402                  30852   
11                  12953                1295304                  31703   
14                  12992                1299206                  32600   
15               

In [16]:
df_top_20_domestic_airports_pivot = df_top_20_domestic_airports.pivot_table(index = ['DEST', 'Route_Type'], values = 'PASSENGERS', aggfunc = 'sum').sort_values('PASSENGERS', ascending = False).reset_index()
df_top_20_domestic_airports_pivot = df_top_20_domestic_airports_pivot.merge(df_top_20_airport_ranks[['DEST', 'Rank']], on = 'DEST')
df_top_20_domestic_airports_pivot.sort_values(['Rank', 'Route_Type'], inplace = True)
df_top_20_domestic_airports_pivot.rename(columns={"DEST":"Airport","PASSENGERS":"Passengers"}, inplace = True)
df_top_20_domestic_airports_pivot

Airport     Route_Type  Passengers  Rank
0      ATL       Domestic  41606158.0     1
1      ATL  International   5782717.0     1
4      LAX       Domestic  29643615.0     2
5      LAX  International  11859213.0     2
2      ORD       Domestic  30105889.0     3
3      ORD  International   6466318.0     3
8      DFW       Domestic  26907771.0     4
9      DFW  International   3915980.0     4
27     JFK       Domestic  14100771.0     5
26     JFK  International  15644170.0     5
6      DEN       Domestic  28045866.0     6
7      DEN  International   1481667.0     6
12     SFO       Domestic  20421699.0     7
13     SFO  International   6553134.0     7
10     LAS       Domestic  21703096.0     8
11     LAS  International   1808553.0     8
18     MCO       Domestic  19933262.0     9
19     MCO  International   3245151.0     9
16     SEA       Domestic  20252954.0    10
17     SEA  International   2305512.0    10
28     EWR       Domestic  15534085.0    11
29     EWR  International   6555068.0    11
14     PHX       Domestic  20320037.0    12
15     PHX  International   1068456.0    12
24     IAH       Domestic  15652841.0    13
25     IAH  International   5151668.0    13
38     MIA       Domestic  10521581.0    14
39     MIA  International  10112517.0    14
20     CLT       Domestic  18536134.0    15
21     CLT  International   1572379.0    15
22     BOS       Domestic  16407792.0    16
23     BOS  International   3517369.0    16
34     FLL       Domestic  13232258.0    17
35     FLL  International   4266110.0    17
30     MSP       Domestic  15527632.0    18
31     MSP  International   1488989.0    18
32     DTW       Domestic  14383588.0    19
33     DTW  International   1683531.0    19
36     PHL       Domestic  12856700.0    20
37     PHL  International   1864367.0    20

I'm also interested in seeing which U.S. airports had the highest international passenger traffic, so I'll create that list as well:

In [17]:
df_top_20_airports_intl_traffic = df_flights_domestic_destination_airports_intl_traffic.pivot_table(index = 'destination_iata_code', values = 'PASSENGERS', aggfunc = 'sum').sort_values('PASSENGERS', ascending = False).head(20) # Calculates the top 20 airports by arriving passenger count.
df_top_20_airports_intl_traffic.reset_index(inplace=True)
df_top_20_airports_intl_traffic['Rank'] = df_top_20_airports_intl_traffic['PASSENGERS'].rank(ascending = False).astype(int) # See https://pandas.pydata.org/docs/reference/api/pandas.Series.rank.html
df_top_20_airports_intl_traffic.rename(columns={"destination_iata_code":"Airport","PASSENGERS":"Passengers"},inplace=True)
df_top_20_airports_intl_traffic

Airport  Passengers  Rank
0      JFK  15644170.0     1
1      LAX  11859213.0     2
2      MIA  10112517.0     3
3      EWR   6555068.0     4
4      SFO   6553134.0     5
5      ORD   6466318.0     6
6      ATL   5782717.0     7
7      IAH   5151668.0     8
8      FLL   4266110.0     9
9      DFW   3915980.0    10
10     IAD   3840026.0    11
11     BOS   3517369.0    12
12     MCO   3245151.0    13
13     SEA   2305512.0    14
14     HNL   2036051.0    15
15     PHL   1864367.0    16
16     LAS   1808553.0    17
17     DTW   1683531.0    18
18     CLT   1572379.0    19
19     MSP   1488989.0    20

Now that I've created these tables, I'll go ahead and export them to both a local .csv file and to my SQL database:

In [18]:
df_top_20_airports_intl_traffic.to_csv('top_20_airports_by_intl_pax_arrivals_2018.csv', index = False)

df_top_20_airports_intl_traffic.to_sql("top_20_airports_by_intl_pax_arrivals_2018",if_exists = 'replace', con = elephantsql_engine, index = False)

20

In [19]:
df_top_20_domestic_airports_pivot.to_csv('top_20_airports_by_pax_arrivals_2018.csv', index = False)

df_top_20_domestic_airports_pivot.to_sql("top_20_airports_by_pax_arrivals_2018",if_exists = 'replace', con = elephantsql_engine, index = False)

40

# Creating an airlines table

I'll now go through a similar process to create a list of the top 20 airlines by US airport arrivals that distinguishes between domestic and international traffic.

In [20]:
df_top_20_airline_ranks = df_flights.pivot_table(index = 'UNIQUE_CARRIER', values = 'PASSENGERS', aggfunc = 'sum').reset_index().sort_values('PASSENGERS', ascending = False).head(20).reset_index(drop=True)
df_top_20_airline_ranks['Rank'] = df_top_20_airline_ranks['PASSENGERS'].rank(ascending = False).astype(int) # See https://pandas.pydata.org/docs/reference/api/pandas.Series.rank.html
df_top_20_airline_ranks

UNIQUE_CARRIER   PASSENGERS  Rank
0              WN  165139267.0     1
1              AA  146267939.0     2
2              DL  143049045.0     3
3              UA  110385926.0     4
4              B6   41635897.0     5
5              AS   33755225.0     6
6              NK   28777722.0     7
7              OO   26810281.0     8
8              F9   19411454.0     9
9              YX   17474019.0    10
10             YV   12590799.0    11
11             9E   11393540.0    12
12             HA   11321558.0    13
13             MQ   10044272.0    14
14             OH    9519918.0    15
15             BA    7457832.0    16
16             EV    7328564.0    17
17             LH    6120312.0    18
18             CP    5869836.0    19
19             WS    5335151.0    20

In [21]:
df_top_20_airline_list = df_top_20_airline_ranks['UNIQUE_CARRIER'].to_list()
df_top_20_airline_list

['WN',
 'AA',
 'DL',
 'UA',
 'B6',
 'AS',
 'NK',
 'OO',
 'F9',
 'YX',
 'YV',
 '9E',
 'HA',
 'MQ',
 'OH',
 'BA',
 'EV',
 'LH',
 'CP',
 'WS']

In [22]:
df_top_20_airlines = df_flights.query("UNIQUE_CARRIER in @df_top_20_airline_list")
df_top_20_airlines

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED     PAYLOAD     SEATS  \
0                        0.0                   1.0     21502.0      76.0   
1                        0.0                   3.0     64506.0     228.0   
2                        0.0                   1.0     21502.0      76.0   
3                        0.0                   1.0     21502.0      76.0   
4                        0.0                   1.0     21502.0      76.0   
...                      ...                   ...         ...       ...   
349158                 855.0                 854.0  27027900.0  109312.0   
349159                 857.0                 859.0  27193800.0  109952.0   
349160                 857.0                 862.0  27292400.0  110336.0   
349163                 930.0                 462.0  14635800.0   59136.0   
349164                 930.0                 464.0  14704600.0   59392.0   

        PASSENGERS   FREIGHT  MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME  \
0              3.0       0.0   0.0     901.0         170.0     140.0   
1             75.0       0.0   0.0     228.0         219.0     140.0   
2             64.0       0.0   0.0     851.0         144.0     114.0   
3             55.0       0.0   0.0     122.0          58.0      31.0   
4              0.0       0.0   0.0     476.0         157.0     138.0   
...            ...       ...   ...       ...           ...       ...   
349158     81674.0  172551.0   0.0     100.0       33015.0   19785.0   
349159     81504.0  479954.0   0.0     100.0       36143.0   19001.0   
349160     85249.0  180613.0   0.0     100.0       33012.0   19820.0   
349163     45376.0   53888.0   0.0     102.0       17165.0    9965.0   
349164     46220.0  208159.0   0.0     102.0       20167.0   10962.0   

       UNIQUE_CARRIER  AIRLINE_ID     UNIQUE_CARRIER_NAME  \
0                  9E     20363.0       Endeavor Air Inc.   
1                  9E     20363.0       Endeavor Air Inc.   
2                  9E     20363.0       Endeavor Air Inc.   
3                  9E     20363.0       Endeavor Air Inc.   
4                  9E     20363.0       Endeavor Air Inc.   
...               ...         ...                     ...   
349158             HA     19690.0  Hawaiian Airlines Inc.   
349159             HA     19690.0  Hawaiian Airlines Inc.   
349160             HA     19690.0  Hawaiian Airlines Inc.   
349163             HA     19690.0  Hawaiian Airlines Inc.   
349164             HA     19690.0  Hawaiian Airlines Inc.   

       UNIQUE_CARRIER_ENTITY REGION CARRIER            CARRIER_NAME  \
0                      07024      D      9E       Endeavor Air Inc.   
1                      07024      D      9E       Endeavor Air Inc.   
2                      07024      D      9E       Endeavor Air Inc.   
3                      07024      D      9E       Endeavor Air Inc.   
4                      07024      D      9E       Endeavor Air Inc.   
...                      ...    ...     ...                     ...   
349158                 05375      D      HA  Hawaiian Airlines Inc.   
349159                 05375      D      HA  Hawaiian Airlines Inc.   
349160                 05375      D      HA  Hawaiian Airlines Inc.   
349163                 05375      D      HA  Hawaiian Airlines Inc.   
349164                 05375      D      HA  Hawaiian Airlines Inc.   

        CARRIER_GROUP  CARRIER_GROUP_NEW  ORIGIN_AIRPORT_ID  \
0                   2                2.0              12264   
1                   2                2.0              12264   
2                   2                2.0              12266   
3                   2                2.0              12323   
4                   2                2.0              12339   
...               ...                ...                ...   
349158              3                3.0              13830   
349159              3                3.0              12173   
349160              3                3.0              13830   
349163        

In [23]:
df_top_20_airlines_pivot = df_top_20_airlines.pivot_table(index = ['UNIQUE_CARRIER', 'Route_Type'], values = 'PASSENGERS', aggfunc = 'sum').sort_values('PASSENGERS', ascending = False).reset_index()
df_top_20_airlines_pivot = df_top_20_airlines_pivot.merge(df_top_20_airline_ranks[['UNIQUE_CARRIER', 'Rank']], on = 'UNIQUE_CARRIER')
df_top_20_airlines_pivot.sort_values(['Rank', 'Route_Type'], inplace = True)
df_top_20_airlines_pivot.rename(columns={"UNIQUE_CARRIER":"Airline","PASSENGERS":"Passengers"}, inplace = True)
df_top_20_airlines_pivot

Airline     Route_Type   Passengers  Rank
0       WN       Domestic  160578652.0     1
1       WN  International    4560615.0     1
4       AA       Domestic  119081636.0     2
5       AA  International   27186303.0     2
2       DL       Domestic  119611221.0     3
3       DL  International   23437824.0     3
6       UA       Domestic   84532551.0     4
7       UA  International   25853375.0     4
8       B6       Domestic   33547937.0     5
9       B6  International    8087960.0     5
10      AS       Domestic   31734733.0     6
11      AS  International    2020492.0     6
12      NK       Domestic   26413644.0     7
13      NK  International    2364078.0     7
14      OO       Domestic   25282225.0     8
15      OO  International    1528056.0     8
16      F9       Domestic   18644380.0     9
17      F9  International     767074.0     9
18      YX       Domestic   16428097.0    10
19      YX  International    1045922.0    10
20      YV       Domestic   11167576.0    11
21      YV  International    1423223.0    11
22      9E       Domestic   10461234.0    12
23      9E  International     932306.0    12
24      HA       Domestic   10319056.0    13
25      HA  International    1002502.0    13
28      MQ       Domestic    8848416.0    14
29      MQ  International    1195856.0    14
26      OH       Domestic    9334599.0    15
27      OH  International     185319.0    15
31      BA       Domestic       7764.0    16
30      BA  International    7450068.0    16
32      EV       Domestic    6493581.0    17
33      EV  International     834983.0    17
35      LH       Domestic       6909.0    18
34      LH  International    6113403.0    18
38      CP       Domestic    5223764.0    19
39      CP  International     646072.0    19
37      WS       Domestic        116.0    20
36      WS  International    5335035.0    20

In [24]:
df_top_20_airlines_pivot.to_csv('top_20_airlines_by_passengers_2018.csv', index = False)
df_top_20_airlines_pivot.to_sql("top_20_airlines_by_passengers_2018",if_exists = 'replace', con = elephantsql_engine, index = False)

40

# Creating a pivot table showing arriving passensger totals for each airline-airport pair:

In [25]:
df_airport_airline_pivot = df_flights.pivot_table(index = ['UNIQUE_CARRIER', 'DEST', "Route_Type"], values = 'PASSENGERS', aggfunc = 'sum').sort_values('PASSENGERS', ascending = False).query("PASSENGERS > 0").reset_index()
df_airport_airline_pivot.rename(columns={"UNIQUE_CARRIER":"Airline","DEST":"Airport","PASSENGERS":"Passengers"},inplace=True)
df_airport_airline_pivot

Airline Airport     Route_Type  Passengers
0         DL     ATL       Domestic  30464398.0
1         AA     DFW       Domestic  19283327.0
2         AA     CLT       Domestic  12510203.0
3         UA     ORD       Domestic  10421303.0
4         WN     MDW       Domestic   9834058.0
...      ...     ...            ...         ...
7926     24Q     SIR  International         1.0
7927     0CQ     PHL       Domestic         1.0
7928     24Q     ATL  International         1.0
7929     0CQ     ACK       Domestic         1.0
7930     VJT     VLC  International         1.0

[7931 rows x 4 columns]

The output of the following two cells should match; otherwise, I've lost some passengers in the process of creating my pivot table.

In [26]:
df_airport_airline_pivot['Passengers'].sum()

951220029.0

In [27]:
df_flights['PASSENGERS'].sum()

951221969.0

In [28]:
df_airport_airline_pivot.to_csv('airport_airline_pairs_2018.csv', index = False)
df_airport_airline_pivot.to_sql('airline_airport_pairs_2018', if_exists = 'replace', con = elephantsql_engine, index = False)

931

# Creating a sample of the data that can be used for an interactive comparison chart:

In order to demonstrate how Plotly can be used to create a chart that lets users choose on which variables to compare data, I'll create a simplified version of the table that contains all of the top 4 airlines' passenger traffic at all of the top 5 airports. This table will also distinguish between domestic and international traffic.

In [29]:
top_5_airports = df_flights_domestic_destination_airports.pivot_table(index = "DEST", values = "PASSENGERS", aggfunc = "sum").sort_values("PASSENGERS", ascending = False).index[0:5].to_list()
top_5_airports

['ATL', 'LAX', 'ORD', 'DFW', 'JFK']

In [30]:
top_4_airlines = df_flights.pivot_table(index = "UNIQUE_CARRIER", values = "PASSENGERS", aggfunc = "sum").sort_values("PASSENGERS", ascending = False).index[0:4].to_list()
top_4_airlines

['WN', 'AA', 'DL', 'UA']

In [31]:
df_flights_excerpt_for_pivot_table = df_flights_domestic_destination_airports.query("UNIQUE_CARRIER in @top_4_airlines & DEST in @top_5_airports")
df_flights_excerpt_for_pivot_table

DEPARTURES_SCHEDULED  DEPARTURES_PERFORMED     PAYLOAD    SEATS  \
3596                     0.0                   1.0    103900.0    273.0   
3597                     0.0                   1.0     50200.0    181.0   
3602                     0.0                   1.0    103900.0    273.0   
3603                     0.0                   1.0     50200.0    181.0   
3606                     0.0                   1.0     50200.0    181.0   
...                      ...                   ...         ...      ...   
348976                 412.0                 411.0  20461760.0  81783.0   
348983                 419.0                 417.0  16732610.0  66840.0   
348991                 425.0                 409.0  16365200.0  65440.0   
349012                 437.0                 431.0  21464760.0  85738.0   
349022                 444.0                 441.0  21967560.0  87756.0   

        PASSENGERS   FREIGHT      MAIL  DISTANCE  RAMP_TO_RAMP  AIR_TIME  \
3596         186.0       0.0       0.0     731.0         137.0     112.0   
3597         134.0       0.0       0.0     731.0         142.0     117.0   
3602         176.0       0.0       0.0     731.0         107.0      90.0   
3603         135.0       0.0       0.0     731.0         120.0      99.0   
3606         161.0       0.0       0.0     594.0         130.0     104.0   
...            ...       ...       ...       ...           ...       ...   
348976     75597.0   93143.0    8766.0     404.0       37671.0   25620.0   
348983     56741.0   14599.0  171876.0     733.0       66352.0   49557.0   
348991     61214.0   11241.0  185047.0     733.0       65540.0   46181.0   
349012     73329.0   94752.0   11778.0     404.0       38217.0   26680.0   
349022     78080.0  110780.0   12645.0     404.0       39316.0   27521.0   

       UNIQUE_CARRIER  AIRLINE_ID     UNIQUE_CARRIER_NAME  \
3596               AA     19805.0  American Airlines Inc.   
3597               AA     19805.0  American Airlines Inc.   
3602               AA     19805.0  American Airlines Inc.   
3603               AA     19805.0  American Airlines Inc.   
3606               AA     19805.0  American Airlines Inc.   
...               ...         ...                     ...   
348976             DL     19790.0    Delta Air Lines Inc.   
348983             AA     19805.0  American Airlines Inc.   
348991             AA     19805.0  American Airlines Inc.   
349012             DL     19790.0    Delta Air Lines Inc.   
349022             DL     19790.0    Delta Air Lines Inc.   

       UNIQUE_CARRIER_ENTITY REGION CARRIER            CARRIER_NAME  \
3596                   0A050      D      AA  American Airlines Inc.   
3597                   0A050      D      AA  American Airlines Inc.   
3602                   0A050      D      AA  American Airlines Inc.   
3603                   0A050      D      AA  American Airlines Inc.   
3606                   0A050      D      AA  American Airlines Inc.   
...                      ...    ...     ...                     ...   
348976                 01260      D      DL    Delta Air Lines Inc.   
348983                 0A050      D      AA  American Airlines Inc.   
348991                 0A050      D      AA  American Airlines Inc.   
349012                 01260      D      DL    Delta Air Lines Inc.   
349022                 01260      D      DL    Delta Air Lines Inc.   

        CARRIER_GROUP  CARRIER_GROUP_NEW  ORIGIN_AIRPORT_ID  \
3596                3                3.0              10397   
3597                3                3.0              10397   
3602                3                3.0              11298   
3603                3                3.0              11298   
3606                3                3.0              13303   
...               ...                ...                ...   
348976              3                3.0              13204   
348983              3                3.0              12953   
348991              3                3.0 

The following pivot table contains rows for each airline/airport/route type pair (e.g. ATL-DL-International; DFW-AA-Domestic, etc.). Since there are 5 airports, 4 airlines, and 2 route types, the pivot table could have up to 40 rows. In practice, however, it will contain a smaller number, since not all of these airlines fly domestically and internationally from every airport.

In [32]:
df_t5_airports_t4_airlines_pivot = df_flights_excerpt_for_pivot_table.pivot_table(index = ["DEST", "UNIQUE_CARRIER", "Route_Type"], values = "PASSENGERS", aggfunc = "sum").reset_index().sort_values(['DEST', 'UNIQUE_CARRIER', 'Route_Type']).reset_index(drop=True)
df_t5_airports_t4_airlines_pivot.rename(columns={"DEST":"Airport","UNIQUE_CARRIER":"Airline","PASSENGERS":"Passengers"},inplace=True)
df_t5_airports_t4_airlines_pivot

Airport Airline     Route_Type  Passengers
0      ATL      AA       Domestic   1271729.0
1      ATL      AA  International       399.0
2      ATL      DL       Domestic  30464398.0
3      ATL      DL  International   4585330.0
4      ATL      UA       Domestic    566751.0
5      ATL      WN       Domestic   5003515.0
6      ATL      WN  International     89823.0
7      DFW      AA       Domestic  19283327.0
8      DFW      AA  International   2530917.0
9      DFW      DL       Domestic    815244.0
10     DFW      DL  International        13.0
11     DFW      UA       Domestic    761181.0
12     DFW      UA  International       604.0
13     DFW      WN       Domestic       234.0
14     JFK      AA       Domestic   2040369.0
15     JFK      AA  International    967172.0
16     JFK      DL       Domestic   4311106.0
17     JFK      DL  International   2392013.0
18     JFK      UA  International       215.0
19     LAX      AA       Domestic   5755386.0
20     LAX      AA  International    859597.0
21     LAX      DL       Domestic   5277140.0
22     LAX      DL  International    685295.0
23     LAX      UA       Domestic   4580527.0
24     LAX      UA  International    521675.0
25     LAX      WN       Domestic   4974955.0
26     LAX      WN  International    125155.0
27     ORD      AA       Domestic   8621345.0
28     ORD      AA  International    743393.0
29     ORD      DL       Domestic   1045269.0
30     ORD      DL  International        47.0
31     ORD      UA       Domestic  10421303.0
32     ORD      UA  International   1732680.0
33     ORD      WN  International       129.0

In [33]:
df_t5_airports_t4_airlines_pivot.to_csv('t5_airports_t4_airlines_2018.csv', index = False)
df_t5_airports_t4_airlines_pivot.to_sql("t5_airports_t4_airlines_2018",if_exists = 'replace', con = elephantsql_engine, index = False)

34